### PR TITLE
wallet: Add ANY_TADDR special string to z_sendmany

### PR DESCRIPTION
**Pull request by str4d**
When using this special string as the from address, non-coinbase UTXOs
from any transparent addresses within the wallet will be used to fund the
transaction. Change outputs will be sent to a new transparent address,
as with any other spend of transparent funds.

Closes zcash/zcash#3640.

### CODE REVIEW

**Comment by str4d** [comment]
@zkbot try

**Comment by str4d** [comment]
The failure above was caused by a CI backend failure, but it [internally retried](https://ci.z.cash/#/builders/9/builds/126), and the only failure there was a pyflakes issue. Pushed to fix.

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -4021,6 +4025,7 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
             "\"operationid\"          (string) An operationid to pass to z_getoperationstatus to get the result of the operation.\n"
             "\nExamples:\n"
             + HelpExampleCli("z_sendmany", "\"t1M72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\" '[{\"address\": \"ztfaW34Gj9FrnGUEf833ywDVL62NWXBM81u6EQnM6VR45eYnXhwztecW1SjxA7JrmAXKJhxhj3vDNEpVCQoSvVoSpmbhtjf\" ,\"amount\": 5.0}]'")
+            + HelpExampleCli("z_sendmany", "\"ANY_TADDR\" '[{\"address\": \"t1M72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\" ,\"amount\": 2.0}]'")
```
Comment (daira) [inline src/wallet/rpcwallet.cpp:4028]:
```suggestion
            + HelpExampleCli("z_sendmany", "\"ANY_TADDR\" '[{\"address\": \"t1M72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\", \"amount\": 2.0}]'")
```

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -4002,12 +4002,16 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
...
+            "\naddress, while change generated from a shielded address returns to itself."
+            "\nWhen sending coinbase UTXOs to a shielded address, change is not allowed."
+            "\nThe entire value of the UTXO(s) must be consumed."
             + strprintf("\nBefore Sapling activates, the maximum number of zaddr outputs is %d due to transaction size limits.\n", Z_SENDMANY_MAX_ZADDR_OUTPUTS_BEFORE_SAPLING)
             + HelpRequiringPassphrase() + "\n"
             "\nArguments:\n"
-            "1. \"fromaddress\"         (string, required) The taddr or zaddr to send the funds from.\n"
+            "1. \"fromaddress\"         (string, required) The transparent or shielded address to send the funds from.\n"
+            "                           The following special strings are also accepted:\n"
+            "                               - \"ANY_TADDR\": Select UTXOs from any transparent addresses belonging to the wallet.\n"
```
Comment (daira) [inline src/wallet/rpcwallet.cpp:4014]:
```suggestion
            "                               - \"ANY_TADDR\": Select non-coinbase UTXOs from any transparent addresses belonging to the wallet.\n"
```

**Review by daira — APPROVED** [review]
utACK with comments.

File: src/wallet/rpcwallet.cpp (hunk 2 above)
Comment (str4d) [reply src/wallet/rpcwallet.cpp:4014]: The PR doesn't currently make this restriction. The current behaviour is the same as for `z_sendmany` with a single t-address: if there is only a single z-address output, then coinbase will be selected. I think however that it's a good suggestion, as anyone wanting to use this with coinbase can just use `z_mergetoaddress`.

**Review by daira — APPROVED** [review]
utACK

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -4002,12 +4002,17 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
...
+            "\nWhen sending coinbase UTXOs to a shielded address, change is not allowed."
+            "\nThe entire value of the UTXO(s) must be consumed."
             + strprintf("\nBefore Sapling activates, the maximum number of zaddr outputs is %d due to transaction size limits.\n", Z_SENDMANY_MAX_ZADDR_OUTPUTS_BEFORE_SAPLING)
             + HelpRequiringPassphrase() + "\n"
             "\nArguments:\n"
-            "1. \"fromaddress\"         (string, required) The taddr or zaddr to send the funds from.\n"
+            "1. \"fromaddress\"         (string, required) The transparent or shielded address to send the funds from.\n"
+            "                           The following special strings are also accepted:\n"
+            "                               - \"ANY_TADDR\": Select non-coinbase UTXOs from any transparent addresses belonging to the wallet.\n"
+            "                                                Use z_shieldcoinbase to shield coinbase UTXOs from multiple transparent addresses.\n"
```
Comment (daira) [inline src/wallet/rpcwallet.cpp:4015]:
```suggestion
            "                                              Use z_shieldcoinbase to shield coinbase UTXOs from multiple transparent addresses.\n"
```

**Comment by daira** [comment]
@zkbot r+

**Comment by str4d** [comment]
[Fixed!](https://github.com/zcash/zcash/compare/78427938caca8e002cffc3980490358ecba4a2c0..3311a85dbe99b6eec1403cfcc232b1d4be51e8d9)

@zkbot r+

**Comment by str4d** [comment]
Transient failure 😑

@zkbot retry

**Pull request merged**